be/src/types/hll_sketch.h

```diff
@@ -0,0 +1,143 @@
+
+    DataSketchesHll(DataSketchesHll&& other) noexcept
+            : _memory_usage(std::move(other._memory_usage)),
+              _sketch_union(std::move(other._sketch_union)),
+              _tgt_type(other._tgt_type) {}
+    DataSketchesHll& operator=(DataSketchesHll&& other) noexcept {
+        if (this != &other) {
+            this->_memory_usage = std::move(other._memory_usage);
+            this->_sketch_union = std::move(other._sketch_union);
+            this->_tgt_type = other._tgt_type;
+        }
+        return *this;
+    }
+
+    explicit DataSketchesHll(const Slice& src, int64_t* memory_usage);
+
+    ~DataSketchesHll() = default;
+
+    // Returns sketch's configured lg_k value.
+    uint8_t get_lg_config_k() const {
+        if (UNLIKELY(_sketch_union == nullptr)) {
+            return DEFAULT_HLL_LOG_K;
+        }
+        return _sketch_union->get_lg_config_k();
+    }
+
+    // Returns the sketch's target HLL mode (from #target_hll_type).
+    datasketches::target_hll_type get_target_type() const {
+        if (UNLIKELY(_sketch_union == nullptr)) {
+            return DEFAULT_HLL_TGT_TYPE;
+        }
+        return _sketch_union->get_target_type();
+    }
+
+    // Add a hash value to this HLL value
+    // NOTE: input must be a hash_value
+    void update(uint64_t hash_value);
+
+    // merge with other HLL value
+    void merge(const DataSketchesHll& other);
+
+    // Return max size of serialized binary
+    size_t max_serialized_size() const;
+    int64_t mem_usage() const { return _memory_usage == nullptr ? 0L : *_memory_usage; }
+
+    // Input slice should have enough capacity for serialize, which
+    // can be got through max_serialized_size(). If insufficient buffer
+    // is given, this will cause process crash.
+    // Return actual size of serialized binary.
+    size_t serialize(uint8_t* dst) const;
+
+    // Now, only empty HLL support this funciton.
+    bool deserialize(const Slice& slice);
+
+    int64_t estimate_cardinality() const;
+
+    // No need to check is_valid for datasketches HLL,
+    // return ture for compatibility.
+    static bool is_valid(const Slice& slice);
+
+    // only for debug
+    std::string to_string() const;
+
+    uint64_t serialize_size() const;
+
+    // common interface
+    void clear() {
+        if (_sketch_union != nullptr) {
+            _sketch_union->reset();
+            _is_changed = true; // Mark as changed after reset
+        }
+    }
+
+    // get hll_sketch object which is lazy initialized
+    hll_sketch_type* get_hll_sketch() const {
+        if (_is_changed) {
+            if (_sketch_union == nullptr) {
+                return nullptr;
+            }
+            _sketch = std::make_unique<hll_sketch_type>(_sketch_union->get_result(_tgt_type));
+            _is_changed = false;
+        }
+        return _sketch.get();
+    }
+
+    inline void mark_changed() { _is_changed = true; }
+
+private:
+    int64_t* _memory_usage;
```

Comment:
why use int64_t* instead of int64_t?

Reply:
`_memory_usage` should bind with `_sketch_union`'s lifecycle. If `_sketch_union` has changed but it pins an old _memory_usage pointer, then `usage after stack return` will meet.

The root cause is that `get_hll_sketch` is called lazily, one `DataSketchesHll` can be created at `HllSketchAggregateFunction #merge` and keeps the stack `memory_usage` and use it in `get_hll_sketch`  but the stack `memory_usage` has been poped.
```
   // get hll_sketch object which is lazy initialized
    hll_sketch_type* get_hll_sketch() const {
        if (_is_changed) {
            if (_sketch_union == nullptr) {
                return nullptr;
            }
            _sketch = std::make_unique<hll_sketch_type>(_sketch_union->get_result(_tgt_type));
            _is_changed = false;
        }
        return _sketch.get();
    }
```

So it's safer to keep `memory_usage`'s lifecycle in the agg state to be as long as `DataSketchesHll` state.